### DISCUSSION TITLE
Best practices for large scrollable areas

### DISCUSSION THREAD

**Discussion by elliottslaughter** [emilk/egui#2443] [Q&A] [answered]
The current version of the egui README says:

> In particular, having a very large UI in a scroll area (with very long scrollback) can be slow, as the content needs to be layed out each frame.

Is there a demo or example that shows how to do scrolling "right"? I understand in principle how to "lay out the part that is in view" for my application, but doesn't this mean essentially building my own entire scrolling implementation in an application-specific way? Or does egui provide hints for me so that I can easily nop-out parts of my UI that are offscreen?

E.g., suppose my application is a set of vertically-stacked boxes of fixed size. How do I allow this to scroll efficiently when the set of boxes grows?

Thanks!

**Comment by tkkcc** [comment, marked as answer]
Check the scrolling and table demo in egui.rs. The requirement is row/cell heights.

**Reply by elliottslaughter** [reply]
Thanks for the response.

My rows are non-uniform (but fixed) height, so I'm look at the "large canvas" example here:

https://github.com/emilk/egui/blob/51081d69fc3ede1d26fd7b722eb9fcaee4660b00/crates/egui_demo_lib/src/demo/scrolling.rs#L118-L121

I developed a basic version from this, but it relies on using the `Painter` API. But in my actual application, each "row" is really a widget and I'll want to do things like make portions that are clickable or detect hover. It's not obvious to me how to do this the "right" way in egui. I could stick with `Painter`, draw everything entirely custom, and maybe find a way to thread the click/hover detection logic through. Or I could move to widgets, but then it's not clear to me how I position these, since I currently rely on absolute positioning with `Painter`.

Any thoughts?

**Reply by elliottslaughter** [reply]
I figured out how to make click targets with `ui.allocate_rect()`, which is enough to get going with the Painter API.

Happy to get any other thoughts on this from the community, but will mark this answered in the meantime since I think I have enough to work with for now.

**Reply by tkkcc** [reply]
Have you check this demo https://github.com/emilk/egui/blob/master/crates/egui_demo_lib/src/demo/table_demo.rs#L204 .
The row heights can be different.

**Reply by elliottslaughter** [reply]
This almost works. But I can't quite figure out how to keep a mutable reference to my table elements without materializing the contents of the height. What I want to write is something like:

```rust
table.body(|mut body| {
    body.heterogeneous_rows(
        self.slots.iter().map(|slot| slot.height()),
        |index, mut row| {
            let slot = &mut self.slots[index];
            // Render slot...
        },
    )
});
```

This fails unless I add `.collect::<Vec<_>>().into_iter()` after `.map(...)`.

It's probably not a huge deal, but it would save an allocation if I could avoid the `collect`.

I assume this is a common pattern, is there a way to deal with this?

Thanks again.
